File: web/src/lib/scoring.ts
export type Answer = 'yes' | 'ifneedbe' | 'no'

export type OptionScore = { yes: number; ifneedbe: number; no: number; score: number }

const POINTS: Record<Answer, number> = { yes: 2, ifneedbe: 1, no: 0 }

export function scoreOptions(
  optionIds: string[],
  votes: { optionId: string; answer: Answer }[],
): Record<string, OptionScore> {
  const result: Record<string, OptionScore> = {}
  for (const id of optionIds) {
    result[id] = { yes: 0, ifneedbe: 0, no: 0, score: 0 }
  }
  for (const vote of votes) {
    const entry = result[vote.optionId]
    if (!entry) continue
    entry[vote.answer] += 1
    entry.score += POINTS[vote.answer]
  }
  return result
}

export function bestOptionId(
  orderedOptionIds: string[],
  scores: Record<string, OptionScore>,
): string | null {
  let bestId: string | null = null
  let bestScore = 0
  for (const id of orderedOptionIds) {
    const score = scores[id]?.score ?? 0
    if (score > bestScore) {
      bestScore = score
      bestId = id
    }
  }
  return bestId
}

const CYCLE: (Answer | null)[] = [null, 'yes', 'ifneedbe', 'no']

export function nextAnswer(current: Answer | null, allowIfNeedBe: boolean): Answer | null {
  const cycle = allowIfNeedBe ? CYCLE : CYCLE.filter((a) => a !== 'ifneedbe')
  const index = cycle.indexOf(current)
  const nextIndex = (index + 1) % cycle.length
  return cycle[nextIndex] ?? null
}

File: internal/polls/scoring.go
package polls

// Ports src/lib/scoring.ts's scoreOptions/bestOptionId — the ranked-choice tally for a
// datetime/options poll (a signup poll never scores; see buildView in service.go).

var answerPoints = map[string]int{"yes": 2, "ifneedbe": 1, "no": 0}

// vote is the minimal shape scoreOptions needs; service.go builds these from queries.Vote rows.
type vote struct {
	optionID string
	answer   string
}

func scoreOptions(optionIDs []string, votes []vote) map[string]OptionScore {
	result := make(map[string]OptionScore, len(optionIDs))
	for _, id := range optionIDs {
		result[id] = OptionScore{}
	}
	for _, v := range votes {
		entry, ok := result[v.optionID]
		if !ok {
			continue
		}
		switch v.answer {
		case "yes":
			entry.Yes++
		case "ifneedbe":
			entry.IfNeedBe++
		case "no":
			entry.No++
		}
		entry.Score += answerPoints[v.answer]
		result[v.optionID] = entry
	}
	return result
}

func bestOptionID(orderedOptionIDs []string, scores map[string]OptionScore) *string {
	var bestID *string
	bestScore := 0
	for _, id := range orderedOptionIDs {
		score := scores[id].Score
		if score > bestScore {
			s := id
			bestScore = score
			bestID = &s
		}
	}
	return bestID
}
